### PR TITLE
Make GlobalPhase not differentiable

### LINKED ISSUE #5541: [BUG] `jax.grad` + `jax.jit` does not work with `AmplitudeEmbedding` and finite shots
### Expected behavior

`qml.AmplitudeEmbedding` should work with `jit` , `grad` and finite shots

```python
from pennylane import numpy as np
import pennylane as qml
import jax 

# with shots = None it works
dev = qml.device("default.qubit", wires = 4, shots=100)

@qml.qnode(dev)
def circuit(coeffs):
        qml.AmplitudeEmbedding(coeffs, normalize = True, wires = [0,1])
        return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

params = jax.numpy.array([0.4, 0.5, 0.1, 0.3])

jac_fn = jax.jacobian(circuit)
# without jax.jit it works
jac_fn = jax.jit(jac_fn)

jac = jac_fn(params)
print(jac)
```

### Actual behavior

`ValueError: need at least one array to stack`

### Additional information

Same issue with `qml.StatePrep` and `qml.MottonenStatePreparation`

### Source code

_No response_

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.36.0.dev0
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /usr/local/lib/python3.10/dist-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane_Lightning

Platform info:           Linux-6.1.58+-x86_64-with-glibc2.35
Python version:          3.10.12
Numpy version:           1.25.2
Scipy version:           1.11.4
Installed devices:
- default.clifford (PennyLane-0.36.0.dev0)
- default.gaussian (PennyLane-0.36.0.dev0)
- default.mixed (PennyLane-0.36.0.dev0)
- default.qubit (PennyLane-0.36.0.dev0)
- default.qubit.autograd (PennyLane-0.36.0.dev0)
- default.qubit.jax (PennyLane-0.36.0.dev0)
- default.qubit.legacy (PennyLane-0.36.0.dev0)
- default.qubit.tf (PennyLane-0.36.0.dev0)
- default.qubit.torch (PennyLane-0.36.0.dev0)
- default.qutrit (PennyLane-0.36.0.dev0)
- null.qubit (PennyLane-0.36.0.dev0)
- lightning.qubit (PennyLane_Lightning-0.35.1)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by Tarun-Kumar07**
**Context:**
When using the following state preparation methods (`AmplitudeEmbedding`, `StatePrep`, `MottonenStatePreparation`) with `jit` and `grad`, the error `ValueError: need at least one array to stack` was encountered.

**Description of the Change:**
All state preparation strategies used `GlobalPhase` under the hood, which caused the above error. After this PR, `GlobalPhase` may not be differentiable anymore, as its `grad_method` is set to `None`.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
It fixes  #5541

### CODE REVIEW

File: pennylane/ops/op_math/controlled.py [hunk 1]
```diff
@@ -837,7 +837,7 @@ def name(self):
 
     @property
     def grad_method(self):
-        return self.base.grad_method
+        return "A" if isinstance(self.base, qml.GlobalPhase) else self.base.grad_method
```
Comment (Tarun-Kumar07) [inline pennylane/ops/op_math/controlled.py:840]: Rather than doing this special handling for `GlobalPhase` should we do this whenever `base.grad_method` is None.
```python
     return "A" if self.base.grad_method is None else self.base.grad_method
```

**Comment by albi3ro** [comment]
Thanks for this @Tarun-Kumar07 

For the failures due to errors:

```
ValueError: Computing the gradient of circuits that return the state with the parameter-shift rule gradient transform is not supported, as it is a hardware-compatible method.
```

That would be expected, and we should shift the measurement to expectation values.

For the failures due to:
```
FAILED tests/templates/test_state_preparations/test_mottonen_state_prep.py::test_jacobian_with_and_without_jit_has_same_output_with_high_shots[StatePrep] - AssertionError: assert Array(False, dtype=bool)
 +  where Array(False, dtype=bool) = <function allclose at 0x7fdb77481940>(Array([-0.0003,  0.0153, -0.0153,  0.0003], dtype=float64), Array([ 1.0187, -0.9953, -1.0047,  0.9813], dtype=float64), atol=0.02)
```

Those are legitimately different results, so we can safely safe we are getting wrong results in that case 😢 I'll investigate.

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -67,9 +67,11 @@
 <h3>Documentation 📝</h3>
 
 <h3>Bug fixes 🐛</h3>
+* Fix `jax.grad` + `jax.jit` not working for `AmplitudeEmbedding`, `StatePrep` and `MottonenStatePreparation`.[(#5620)](https://github.com/PennyLaneAI/pennylane/pull/5620)
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:70]:
```suggestion
* Fix `jax.grad` + `jax.jit` not working for `AmplitudeEmbedding`, `StatePrep` and `MottonenStatePreparation`.
  [(#5620)](https://github.com/PennyLaneAI/pennylane/pull/5620) 
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -67,9 +67,11 @@
 <h3>Documentation 📝</h3>
 
 <h3>Bug fixes 🐛</h3>
+* Fix `jax.grad` + `jax.jit` not working for `AmplitudeEmbedding`, `StatePrep` and `MottonenStatePreparation`.[(#5620)](https://github.com/PennyLaneAI/pennylane/pull/5620) 
 
 <h3>Contributors ✍️</h3>
 
 This release contains contributions from (in alphabetical order):
+Tarun Kumar Allamsetty
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:75]:
```suggestion
Tarun Kumar Allamsetty,
```

File: pennylane/ops/identity.py [hunk 4]
```diff
@@ -281,7 +281,7 @@ def circuit():
 
     """
 
-    grad_method = "A"
+    grad_method = None
```
Comment (dwierichs) [inline pennylane/ops/identity.py:284]: May I suggest to put `grad_method="F"` here? Setting it to `None` has the disadvantage that `param_shift` will think it can "tackle" a global phase with `return qml.state()`, because it thinks there is no differentiation needed. With `grad_method="F"`, `finite-diff` and `param_shift` will behave as expected, computing the state derivative correctly and raising an error (because param-shift+state does not work), respectively.
The only downside I currently see for `grad_method="F"` is that `finite-diff` (and `param_shift` for valid measurements, because it has a fallback logic) will create tapes to differentiate global phases even when they have no impact (i.e. expectation values, probabilities etc).

However, I'd prefer that we are wasteful (unnecessary global phases can be compiled away anyways, and then some form of caching should kick in) than returning wrong results.

File: pennylane/ops/op_math/controlled.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/ops/op_math/controlled.py:840]: Could keep this the same even if `GlobalPhase.grad_method = "F"`.

File: tests/templates/test_embeddings/test_amplitude.py [hunk 5]
```diff
@@ -453,3 +453,47 @@ def test_torch(self, tol, features, pad_with, normalize):
         res2 = circuit2(features, pad_with)
 
         assert qml.math.allclose(res, res2, atol=tol, rtol=0)
+
+
+@pytest.mark.jax
+def test_jacobian_with_and_without_jit_has_same_output_with_high_shots():
+    import jax
+
+    dev = qml.device("default.qubit", wires=4, shots=10000, seed=7890234)
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:462]:
```suggestion
    dev = qml.device("default.qubit", shots=10000, seed=7890234)
```

File: tests/templates/test_embeddings/test_amplitude.py [hunk 6]
```diff
@@ -453,3 +453,47 @@ def test_torch(self, tol, features, pad_with, normalize):
...
+
+    jac = jac_fn(params)
+
+    jac_jit = jac_jit_fn(params)
+
+    assert qml.math.allclose(jac, jac_jit, atol=0.02)
+
+
+@pytest.mark.jax
+def test_jacobian_with_and_without_jit_has_same_output_with_analytic_mode():
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:481]: Could combine this test with the one above, but that's a matter of taste. If not, the docstrings of the two tests (also in the next file) should be somewhat different in order to not rely on the test function name alone :)

Comment (dwierichs) [reply tests/templates/test_embeddings/test_amplitude.py:481]: (using `pytest.mark.parametrize`)

File: tests/templates/test_embeddings/test_amplitude.py [hunk 7]
```diff
@@ -453,3 +453,47 @@ def test_torch(self, tol, features, pad_with, normalize):
...
+
+    params = jax.numpy.array([0.4, 0.5, 0.1, 0.3])
+    jac_fn = jax.jacobian(circuit)
+    jac_jit_fn = jax.jit(jac_fn)
+
+    jac = jac_fn(params)
+
+    jac_jit = jac_jit_fn(params)
+
+    assert qml.math.allclose(jac, jac_jit, atol=0.02)
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:499]: Standard `atol` should work here, I think?
```suggestion
    assert qml.math.allclose(jac, jac_jit)
```

File: tests/templates/test_embeddings/test_amplitude.py [hunk 8]
```diff
@@ -453,3 +453,47 @@ def test_torch(self, tol, features, pad_with, normalize):
...
+
+
+@pytest.mark.jax
+def test_jacobian_with_and_without_jit_has_same_output_with_analytic_mode():
+    import jax
+
+    dev = qml.device("default.qubit", wires=4, shots=None)
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:484]:
```suggestion
    dev = qml.device("default.qubit", shots=None)
```

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 9]
```diff
@@ -492,3 +492,51 @@ def circuit(state):
...
     assert qml.math.allclose(actual, expected)
+
+
+@pytest.mark.jax
+@pytest.mark.parametrize("state_prep_op", [qml.StatePrep, qml.MottonenStatePreparation])
+def test_jacobian_with_and_without_jit_has_same_output_with_high_shots(state_prep_op):
+    """Test that the jacobian of the circuit is the same with and without jit"""
+    import jax
+
+    dev = qml.device("default.qubit", wires=4, shots=10000, seed=7890234)
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:503]:
```suggestion
    dev = qml.device("default.qubit", shots=10000, seed=7890234)
```

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 10]
```diff
@@ -492,3 +492,51 @@ def circuit(state):
...
+    assert qml.math.allclose(jac, jac_jit, atol=0.02)
+
+
+@pytest.mark.jax
+@pytest.mark.parametrize("state_prep_op", [qml.StatePrep, qml.MottonenStatePreparation])
+def test_jacobian_with_and_without_jit_has_same_output_with_analytic_mode(state_prep_op):
+    """Test that the jacobian of the circuit is the same with and without jit."""
+    import jax
+
+    dev = qml.device("default.qubit", wires=4, shots=None)
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:527]:
```suggestion
    dev = qml.device("default.qubit", shots=None)
```

File: tests/templates/test_embeddings/test_amplitude.py [hunk 11]
```diff
@@ -453,3 +453,47 @@ def test_torch(self, tol, features, pad_with, normalize):
         res2 = circuit2(features, pad_with)
 
         assert qml.math.allclose(res, res2, atol=tol, rtol=0)
+
+
+@pytest.mark.jax
+def test_jacobian_with_and_without_jit_has_same_output_with_high_shots():
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:459]: Missing docstring :upside_down_face: 
```suggestion
def test_jacobian_with_and_without_jit_has_same_output_with_high_shots():
    """Test that the jacobian of AmplitudeEmbedding is the same with and without jit"""
```

File: tests/templates/test_embeddings/test_amplitude.py (hunk 6 above)
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:481]:
```suggestion
def test_jacobian_with_and_without_jit_has_same_output_with_analytic_mode():
    """Test that the jacobian of AmplitudeEmbedding is the same with and without jit"""
```

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 12]
```diff
@@ -492,3 +492,51 @@ def circuit(state):
...
+
+    params = jax.numpy.array([0.5, 0.5, 0.5, 0.5])
+    jac_fn = jax.jacobian(circuit)
+    jac_jit_fn = jax.jit(jac_fn)
+
+    jac = jac_fn(params)
+
+    jac_jit = jac_jit_fn(params)
+
+    assert qml.math.allclose(jac, jac_jit, atol=0.02)
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:542]:
```suggestion
    assert qml.math.allclose(jac, jac_jit)
```

**Review by dwierichs — CHANGES_REQUESTED** [review]
I left a couple of small comments and one major suggestion: Could we set `GlobalPhase.grad_method = "F"`? This will produce unnecessary shifted tapes for expectation values and probabilities, but it will avoid wrong results when differentiating `qml.state` with `finite_diff` and `param_shift`.

**Comment by dwierichs** [comment]
>Those are legitimately different results, so we can safely safe we are getting wrong results in that case 😢 I'll investigate.

@Tarun-Kumar07 @albi3ro Not sure you got to this yet, but it seems that the decomposition of those state preparation methods handle special parameter values differently than others. This makes the derivative wrong at those special values, because `param_shift` is handed a tape that does not contain the general decomposition, so it will not shift all operations that need shifting. As JITting does not allow such special cases, we only make this mistake without JITting, hence the difference in the results within those tests. 

Basically, the decomposition does something like the following decomposition for `RZ`:
```python
def compute_decomposition(theta, wires):
    if not qml.math.is_abstract(theta) and qml.math.isclose(theta, 0):
        return []
    return [qml.RZ(theta, wires)]
```
It's correct but it does not have the correct parameter-shift derivative at 0.

This looks like an independent bug to me, and like one that could be hiding across the codebase for other ops as well, theoretically.

File: tests/templates/test_embeddings/test_amplitude.py (hunk 6 above)
Comment (dwierichs) [reply tests/templates/test_embeddings/test_amplitude.py:483]: It's not so important, but one could use a joint parametrization
```python
@pytest.mark.parametrize("shots, atol", [(None, 1e-8), (10000, 0.02)])
```
the seed can be set for the `shots=None` device as well afaik, and the `diff_method` actually is `"parameter-shift"` in both cases, just that it has to be set manually for `shots=None` and is obtained by the internal logic for `shots!=None` :) 

But again, this is not terribly important, was just a suggestion.

Comment (Tarun-Kumar07) [reply tests/templates/test_embeddings/test_amplitude.py:483]: I see, thank you for the explanation :)

**Comment by dwierichs** [comment]
@Tarun-Kumar07 Sorry for taking so long with this! We decided to move ahead with this PR as you originally drafted it (with `grad_method=None`). Before this PR can be merged, the fix in #5774 will need to be merged in, though.
I took the liberty of using a test you wrote here for that PR, your work is much appreciated! :pray:

**Comment by Tarun-Kumar07** [comment]
Hey @dwierichs , once the PR #5774 is merged I will revert changes to `grad_method=None`. Thanks for updating :)

**Comment by dwierichs** [comment]
@Tarun-Kumar07 It is merged :) 
I took the liberty to merge the master branch into your branch, updating the tests you wrote (and that made it into #5774) accordingly.

**Comment by dwierichs** [comment]
Hi @Tarun-Kumar07 ,
Do you know when you might have time to get back to this? :) Or would you prefer us to finalize it?

**Comment by Tarun-Kumar07** [comment]
Hi @dwierichs,

I am currently tied up until August 17th. After that, I will be able to work on this.
If this is a priority, I suggest assigning someone else to take over in the meantime.

Thank you for understanding.

**Comment by dwierichs** [comment]
Thanks @Tarun-Kumar07.
That is very understandable of course! I'll try and push your PR through, as it is closed to finish anyways, I think.
Thank you for your contribution, looking forward to the next one :star_struck:

File: tests/templates/test_state_preparations/test_mottonen_state_prep.py [hunk 13]
```diff
@@ -430,16 +430,21 @@ def circuit(coeffs):
...
-    jac_jit = jac_jit_fn(params)
+    jac_fd = jac_fd_fn(params)
+    jac_fd_jit = jac_fd_fn_jit(params)
+    jac_ps = jac_ps_fn(params)
+    jac_ps_jit = jac_ps_fn_jit(params)
 
-    assert qml.math.allclose(jac_exact, jac_jit, atol=atol)
-    assert qml.math.allclose(jac, jac_jit, atol=atol)
+    for compare in [jac_fd, jac_fd_jit, jac_ps, jac_ps_jit]:
+        assert qml.math.allclose(jac_exact, compare, atol=atol)
```
Comment (dwierichs) [inline tests/templates/test_state_preparations/test_mottonen_state_prep.py:450]: Also check that parameter-shift rule, now that it is working, agrees on derivative of MottonenStatePreparation.

File: tests/templates/test_embeddings/test_amplitude.py [hunk 14]
```diff
@@ -584,3 +584,28 @@ def test_torch(self, tol, features, pad_with, dtype):
...
+
+    params = jax.numpy.array([0.4, 0.5, 0.1, 0.3])
+    jac_fn = jax.jacobian(circuit)
+    jac_jit_fn = jax.jit(jac_fn)
+
+    jac = jac_fn(params)
+
+    jac_jit = jac_jit_fn(params)
+
+    assert qml.math.allclose(jac, jac_jit, atol=atol)
```
Comment (dwierichs) [inline tests/templates/test_embeddings/test_amplitude.py:611]: `AmplitudeEmbedding` uses `GlobalPhase` and was the original failing case.

File: tests/ops/qubit/test_parametric_ops.py [hunk 15]
```diff
@@ -2015,6 +2015,7 @@ def test_globalphase_autograd_grad(self, tol, dev_name, diff_method, wires):
         @qml.qnode(dev, diff_method=diff_method)
         def circuit(x):
             qml.Identity(wires[0])
+            qml.GlobalPhase(x, wires=[0, 1])  # Does not change the derivative, but tests it
```
Comment (dwierichs) [inline tests/ops/qubit/test_parametric_ops.py:2018]: Test `GlobalPhase` derivative directly, without drawing in `AmplitudeEmbedding`.

**Review by dwierichs — APPROVED** [review]
Approving the first part of this PR. I modified it slightly, which will need approval by someone else.

**Review by mudit2812 — APPROVED** [review]

File: doc/releases/changelog-dev.md [hunk 16]
```diff
@@ -118,6 +118,11 @@
 
 <h3>Breaking changes 💔</h3>
 
+* `GlobalPhase` is considered non-differentiable with tape transforms.
+  As a consequence, `qml.gradients.finite_diff` and `qml.gradients.spsa_grad` no longer
+  support differentiation of `GlobalPhase`.
+  [(#5620)](https://github.com/PennyLaneAI/pennylane/pull/5620)
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md:124]: @trbromley This is the downside of this PR. I think we can live with that?

Comment (trbromley) [reply doc/releases/changelog-dev.md:143]: :thinking: what causes `finite_diff` to be problematic here but `param_shift` is not?

Comment (trbromley) [reply doc/releases/changelog-dev.md:143]:
```suggestion
* `GlobalPhase` is considered non-differentiable with tape transforms.
  As a consequence, `qml.gradients.finite_diff` and `qml.gradients.spsa_grad` no longer
  support differentiation of `GlobalPhase` with state-based outputs.
  [(#5620)](https://github.com/PennyLaneAI/pennylane/pull/5620) 
```
So should the changelog read like this?

Comment (dwierichs) [reply doc/releases/changelog-dev.md:143]: Oh yeah, that is a fair point! Thank you!

Comment (dwierichs) [reply doc/releases/changelog-dev.md]: @trbromley Are we good to merge, then? :)

**Comment by Alex-Preciado** [comment]
Thank you so much for this contribution @Tarun-Kumar07!! 🚀
